src/tga/encoder.rs

```diff
@@ -123,3 +77,141 @@ impl<W: Write> ImageEncoder for TgaEncoder<W> {
         self.encode(buf, width, height, color_type)
     }
 }
+
+fn get_pixel_info(color_type: ColorType) -> ImageResult<(u8, u8, ImageType)> {
+    let sizes = match color_type {
+        ColorType::Rgba8 | ColorType::Bgra8 => (8, 24, ImageType::RawTrueColor),
+        ColorType::Rgb8 | ColorType::Bgr8 => (0, 24, ImageType::RawTrueColor),
+        ColorType::La8 => (8, 8, ImageType::RawGrayScale),
+        ColorType::L8 => (0, 8, ImageType::RawGrayScale),
+        _ => {
+            return Err(ImageError::Unsupported(
+                UnsupportedError::from_format_and_kind(
+                    ImageFormat::Tga.into(),
+                    UnsupportedErrorKind::Color(color_type.into()),
+                ),
+            ))
+        }
+    };
+
+    Ok(sizes)
+}
+
+fn tga_dimensions(width: u32, height: u32) -> ImageResult<(u16, u16)> {
+    fn inner_dimensions(width: u32, height: u32) -> Option<(u16, u16)> {
+        let width = u16::try_from(width).ok()?;
+        let height = u16::try_from(height).ok()?;
+        Some((width, height))
+    }
+
+    inner_dimensions(width, height).ok_or(ImageError::Parameter(ParameterError::from_kind(
+        ParameterErrorKind::DimensionMismatch,
+    )))
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::super::TgaDecoder;
+    use super::TgaEncoder;
+    use crate::color::ColorType;
+    use crate::image::ImageDecoder;
+    use std::io::Cursor;
+
+    fn round_trip_image(image: &[u8], width: u32, height: u32, c: ColorType) -> Vec<u8> {
+        let mut encoded_data = Vec::new();
+        {
+            let encoder = TgaEncoder::new(&mut encoded_data);
+            encoder
+                .encode(&image, width, height, c)
+                .expect("could not encode image");
+        }
+
+        let decoder = TgaDecoder::new(Cursor::new(&encoded_data)).expect("failed to decode");
+
+        let mut buf = vec![0; decoder.total_bytes() as usize];
+        decoder.read_image(&mut buf).expect("failed to decode");
+        buf
+    }
+
+    // FIXME: Taking a really long time to run. Tried switching to Bgr8, but then it just panics.
+    // #[test]
+    // fn huge_files_return_error() {
+    //     let mut encoded_data = Vec::new();
+    //     let image = vec![0u8; 3 * 40_000 * 40_000]; // 40_000x40_000 pixels, 3 bytes per pixel, allocated on the heap
```

Comment:
Some other tests for dimension errors test them individually. So you'd check that `1×65536` errors and then that `65536×1` errors. This avoids the extremely large single allocation.

Reply:
Yeah, I came to that conclusion as well. As for the slowness, I've determined that it's the byte swap step that occurs when encoding an Rgb8 or Rgba8 image. Is there a faster way to do that operation?

Reply:
I can't say for sure. The binary produced isn't exactly optimal but I can't find an easy way to produce something more optimal quickly either.